### ISSUE TITLE
用example_layers下面的solver训练loss从开始就一直在0.2

### ISSUE THREAD

**Issue by yzk0281** [pinguo-luhaofang/tripletloss#4]
如题 
我用webface-clean进行训练 
用example_layers下面的solver训练loss从开始就一直在0.2 也就是说一开始就收敛了。。
不知道您是怎么训练的？

**Comment by pinguo-luhaofang** [comment]
你确定你的layer的参数初始化是ok的吗？
同时magin设置为0.2 loss也是0.2 这个收敛了吗？

**Comment by yzk0281** [comment]
我没有更改你的网络设置包括：tripletloss-master/example_layer/solver.prototxt
和 
tripletloss-master/example_layer/train.prototxt
除了train.prototxt里面的训练输入:
name: "VGG_FACE_16_layers"
layer {
  name: "data"
  type: "Data"
  top: "data"
  include {
    phase: TRAIN
  }
  transform_param {
    mirror: true
    crop_size: 128
    mean_file: "/root/cnn/caffe-windows-master/data/webface_clean/vgg/lmdb128/train/mean.binaryproto"
  }
  data_param {
    source: "/root/cnn/caffe-windows-master/data/webface_clean/vgg/lmdb128/train"
    batch_size: 30
    backend: LMDB
  }
}
layer {
  name: "data"
  type: "Data"
  top: "data"
  include {
    phase: TEST
  }
  transform_param {
    mirror: false
    crop_size: 128
    mean_file: "/root/cnn/caffe-windows-master/data/webface_clean/vgg/lmdb128/val/mean.binaryproto"
  }
  data_param {
    source: "/root/cnn/caffe-windows-master/data/webface_clean/vgg/lmdb128/val"
    batch_size: 30
    backend: LMDB
  }
}

**Comment by pinguo-luhaofang** [comment]
我不建议固定输入样本的顺序。
样例中，我给出的是随机挑选样本图片构建minibatch的。

**Comment by pinguo-luhaofang** [comment]
loss一开始都是0.2 说明网络中的参数要么开始太小了，学习率太大，造成一开始训练有的参数就为0了；要么就是一开始你的某些layer参数为0，没有初始化，建议输出特定layer的forward，看看

**Comment by yzk0281** [comment]
噢 ~~请问你的vggface的模型重新训练出来了吗？

**Comment by yzk0281** [comment]
你那个是 finetune还是重新训练？

**Comment by yzk0281** [comment]
hello
我准备用官方放出来的vggface model加上你的triploss来finetune能帮忙看一下train.prototxt吗？：
name: "VGG_FACE_16_layers"
layer {
  name: "data"
  type: "Data"
  top: "data"
  include {
    phase: TRAIN
  }
  transform_param {
    mirror: true
    crop_size: 224 
    mean_file: "/root/cnn/caffe-windows-master/data/webface_clean/vgg/lmdb224/train/mean.binaryproto"
  }
  data_param {
    source: "/root/cnn/caffe-windows-master/data/webface_clean/vgg/lmdb224/train"
    batch_size: 30
    backend: LMDB
  }
}
layer {
  name: "data"
  type: "Data"
  top: "data"
  include {
    phase: TEST
  }
  transform_param {
    mirror: false
    crop_size: 224 
    mean_file: "/root/cnn/caffe-windows-master/data/webface_clean/vgg/lmdb224/val/mean.binaryproto"
  }
  data_param {
    source: "/root/cnn/caffe-windows-master/data/webface_clean/vgg/lmdb224/val"
    batch_size: 2 
    backend: LMDB
  }
}
layer {
  name: "conv1_1"
  type: "Convolution"
  bottom: "data"
  top: "conv1_1"
  param {
    lr_mult: 0
    decay_mult: 0
  }
  param {
    lr_mult: 0
    decay_mult: 0
  }
  convolution_param {
    num_output: 64
    pad: 1
    kernel_size: 3
    weight_filler {
      type: "gaussian"
      std: 0.01
    }
  }
}
layer {
  name: "relu1_1"
  type: "ReLU"
  bottom: "conv1_1"
  top: "conv1_1"
}
layer {
  name: "conv1_2"
  type: "Convolution"
  bottom: "conv1_1"
  top: "conv1_2"
  param {
    lr_mult: 0
    decay_mult: 0
  }
  param {
    lr_mult: 0
    decay_mult: 0
  }
  convolution_param {
    num_output: 64
    pad: 1
    kernel_size: 3
    weight_filler {
      type: "gaussian"
      std: 0.01
    }
  }
}
layer {
  name: "relu1_2"
  type: "ReLU"
  bottom: "conv1_2"
  top: "conv1_2"
}
layer {
  name: "pool1"
  type: "Pooling"
  bottom: "conv1_2"
  top: "pool1"
  pooling_param {
    pool: MAX
    kernel_size: 2
    stride: 2
  }
}
layer {
  name: "conv2_1"
  type: "Convolution"
  bottom: "pool1"
  top: "conv2_1"
  param {
    lr_mult: 0
    decay_mult: 0
  }
  param {
    lr_mult: 0
    decay_mult: 0
  }
  convolution_param {
    num_output: 128
    pad: 1
    kernel_size: 3
    weight_filler {
      type: "gaussian"
      std: 0.01
    }
  }
}
layer {
  name: "relu2_1"
  type: "ReLU"
  bottom: "conv2_1"
  top: "conv2_1"
}
layer {
  name: "conv2_2"
  type: "Convolution"
  bottom: "conv2_1"
  top: "conv2_2"
  param {
    lr_mult: 0
    decay_mult: 0
  }
  param {
    lr_mult: 0
    decay_mult: 0
  }
  convolution_param {
    num_output: 128
    pad: 1
    kernel_size: 3
    weight_filler {
      type: "gaussian"
      std: 0.01
    }
  }
}
layer {
  name: "relu2_2"
  type: "ReLU"
  bottom: "conv2_2"
  top: "conv2_2"
}
layer {
  name: "pool2"
  type: "Pooling"
  bottom: "conv2_2"
  top: "pool2"
  pooling_param {
    pool: MAX
    kernel_size: 2
    stride: 2
  }
}
layer {
  name: "conv3_1"
  type: "Convolution"
  bottom: "pool2"
  top: "conv3_1"
  param {
    lr_mult: 1
    decay_mult: 1
  }
  param {
    lr_mult: 2
    decay_mult: 0
  }
  convolution_param {
    num_output: 256
    pad: 1
    kernel_size: 3
    weight_filler {
      type: "gaussian"
      std: 0.01
    }
  }
}
layer {
  name: "relu3_1"
  type: "ReLU"
  bottom: "conv3_1"
  top: "conv3_1"
}
layer {
  name: "conv3_2"
  type: "Convolution"
  bottom: "conv3_1"
  top: "conv3_2"
  param {
    lr_mult: 1
    decay_mult: 1
  }
  param {
    lr_mult: 2
    decay_mult: 0
  }
  convolution_param {
    num_output: 256
    pad: 1
    kernel_size: 3
    weight_filler {
      type: "gaussian"
      std: 0.01
    }
  }
}
layer {
  name: "relu3_2"
  type: "ReLU"
  bottom: "conv3_2"
  top: "conv3_2"
}
layer {
  name: "conv3_3"
  type: "Convolution"
  bottom: "conv3_2"
  top: "conv3_3"
  param {
    lr_mult: 1
    decay_mult: 1
  }
  param {
    lr_mult: 2
    decay_mult: 0
  }
  convolution_param {
    num_output: 256
    pad: 1
    kernel_size: 3
    weight_filler {
      type: "gaussian"
      std: 0.01
    }
  }
}
layer {
  name: "relu3_3"
  type: "ReLU"
  bottom: "conv3_3"
  top: "conv3_3"
}
layer {
  name: "pool3"
  type: "Pooling"
  bottom: "conv3_3"
  top: "pool3"
  pooling_param {
    pool: MAX
    kernel_size: 2
    stride: 2
  }
}
layer {
  name: "conv4_1"
  type: "Convolution"
  bottom: "pool3"
  top: "conv4_1"
  param {
    lr_mult: 1
    decay_mult: 1
  }
  param {
    lr_mult: 2
    decay_mult: 0
  }
  convolution_param {
    num_output: 512
    pad: 1
    kernel_size: 3
    weight_filler {
      type: "gaussian"
      std: 0.01
    }
  }
}
layer {
  name: "relu4_1"
  type: "ReLU"
  bottom: "conv4_1"
  top: "conv4_1"
}
layer {
  name: "conv4_2"
  type: "Convolution"
  bottom: "conv4_1"
  top: "conv4_2"
  param {
    lr_mult: 1
    decay_mult: 1101
  }
  param {
    lr_mult: 2
    decay_mult: 0
  }
  convolution_param {
    num_output: 512
    pad: 1
    kernel_size: 3
    weight_filler {
      type: "gaussian"
      std: 0.01
    }
  }
}
layer {
  name: "relu4_2"
  type: "ReLU"
  bottom: "conv4_2"
  top: "conv4_2"
}
layer {
  name: "conv4_3"
  type: "Convolution"
  bottom: "conv4_2"
  top: "conv4_3"
  param {
    lr_mult: 1
    decay_mult: 1
  }
  param {
    lr_mult: 2
    decay_mult: 0
  }
  convolution_param {
    num_output: 512
    pad: 1
    kernel_size: 3
    weight_filler {
      type: "gaussian"
      std: 0.01
    }
  }
}
layer {
  name: "relu4_3"
  type: "ReLU"
  bottom: "conv4_3"
  top: "conv4_3"
}
layer {
  name: "pool4"
  type: "Pooling"
  bottom: "conv4_3"
  top: "pool4"
  pooling_param {
    pool: MAX
    kernel_size: 2
    stride: 2
  }
}
layer {
  name: "conv5_1"
  type: "Convolution"
  bottom: "pool4"
  top: "conv5_1"
  param {
    lr_mult: 1
    decay_mult: 1
  }
  param {
    lr_mult: 2
    decay_mult: 0
  }
  convolution_param {
    num_output: 512
    pad: 1
    kernel_size: 3
    weight_filler {
      type: "gaussian"
      std: 0.01
    }
  }
}
layer {
  name: "relu5_1"
  type: "ReLU"
  bottom: "conv5_1"
  top: "conv5_1"
}
layer {
  name: "conv5_2"
  type: "Convolution"
  bottom: "conv5_1"
  top: "conv5_2"
  param {
    lr_mult: 1
    decay_mult: 1
  }
  param {
    lr_mult: 2
    decay_mult: 0
  }
  convolution_param {
    num_output: 512
    pad: 1
    kernel_size: 3
    weight_filler {
      type: "gaussian"
      std: 0.01
    }
  }
}
layer {
  name: "relu5_2"
  type: "ReLU"
  bottom: "conv5_2"
  top: "conv5_2"
}
layer {
  name: "conv5_3"
  type: "Convolution"
  bottom: "conv5_2"
  top: "conv5_3"
  param {
    lr_mult: 1
    decay_mult: 1
  }
  param {
    lr_mult: 2
    decay_mult: 0
  }
  convolution_param {
    num_output: 512
    pad: 1
    kernel_size: 3
    weight_filler {
      type: "gaussian"
      std: 0.01
    }
  }
}
layer {
  name: "relu5_3"
  type: "ReLU"
  bottom: "conv5_3"
  top: "conv5_3"
}
layer {
  name: "pool5"
  type: "Pooling"
  bottom: "conv5_3"
  top: "pool5"
  pooling_param {
    pool: MAX
    kernel_size: 2
    stride: 2
  }
}
layer {
  name: "fc6"
  type: "InnerProduct"
  bottom: "pool5"
  top: "fc6"
  param {
    lr_mult: 1
    decay_mult: 1
  }
  param {
    lr_mult: 2
    decay_mult: 0
  }
  inner_product_param {
    num_output: 4096
    weight_filler {
      type: "gaussian"
      std: 0.5
    }
    bias_filler {
      type: "constant"
      value: 0.1
    }
  }
}
layer {
  name: "relu6"
  type: "ReLU"
  bottom: "fc6"
  top: "fc6"
}
layer {
  name: "drop6"
  type: "Dropout"
  bottom: "fc6"
  top: "fc6"
  dropout_param {
    dropout_ratio: 0.5
  }
}
layer {
  name: "fc7"
  type: "InnerProduct"
  bottom: "fc6"
  top: "fc7"
  param {
    lr_mult: 1
    decay_mult: 1
  }
  param {
    lr_mult: 2
    decay_mult: 0
  }
  inner_product_param {
    num_output: 4096
    weight_filler {
      type: "gaussian"
      std: 0.5
    }
    bias_filler {
      type: "constant"
      value: 0.1
    }
  }
}
layer {
  name: "relu7"
  type: "ReLU"
  bottom: "fc7"
  top: "fc7"
}
layer {
  name: "drop7"
  type: "Dropout"
  bottom: "fc7"
  top: "fc7"
  dropout_param {
    dropout_ratio: 0.5
  }
}
layer {
  name: "fc9"
  type: "InnerProduct"
  bottom: "fc7"
  top: "fc9"
  param {
    lr_mult: 1
    decay_mult: 1
  }
  param {
    lr_mult: 2
    decay_mult: 0
  }
  inner_product_param {
    num_output: 10575 
    weight_filler {
      type: "gaussian"
      std: 0.1
    }
    bias_filler {
      type: "constant"
      value: 0
    }
  }
}
layer {
  name: "relu8"
  type: "ReLU"
  bottom: "fc9"
  top: "fc9"
}
layer {
  name: "drop8"
  type: "Dropout"
  bottom: "fc9"
  top: "fc9"
  dropout_param {
    dropout_ratio: 0.5
  }
}
layer {
  name: "norm2"
  type: "Python"
  bottom: "fc9"
  top: "norm2"
  python_param {
    module: "norm2layer"
    layer: "Norm2Layer"
  }
}
layer {
  name: "loss_distance"
  type: "Python"
  bottom: "norm2"
  top: "loss_distance"
  python_param {
    module: "tripletlayer"
    layer: "TripletLayer"
    param_str: "'margin': 0.2"
  }
  loss_weight: 1
}

**Comment by yzk0281** [comment]
用你的train.py提示
assert bottom[0].num % 3 == 0

**Comment by pinguo-luhaofang** [comment]
example_layer中给出的是tripletloss的layer。用到的是offline的一种样本筛选的方法。
tripletloss文件夹中给出来了样本筛选的datalayer的实现以及具体抽样策略，如果你的folder在python找得到的话，通过运行train.py是可以直接训练的。
online的样本抽取策略中，我并没有把文件转成lmdb格式，而是通过image的文件路径对输入样本进行筛选。具体请参见 tripletselectlayer.py和sampledata.py

网络模型是可用的（我就是用的这个网络），但是初始化参数需要在训练的时候进行一些微调。当然如果你是finetune的vgg_face那么可以忽略该条。

**Comment by yzk0281** [comment]
Hi,
我如果用你的datalayer和online策略进行finetune的话。。
config.py里面的配置我想请教下 
IMAGEPATH下面的图片是不是要按人名（或者分类）来分文件夹存放？
SAMPLEPATH是不是:
文件名  人名(类)
格式的txt文件 
SAMPLEPATH中包含所有图片还是部分（包含在训练集中）？

**Comment by pinguo-luhaofang** [comment]
最好是先调整网络结构做分类的训练，然后通过分类训练的参数继续进行triplet的finetune。
IMAGEPATH是文件的存放文件夹地址。
SAMPLEPATH是txt文件，保存的是所有样本文件的名称和类别。
我的文件命名规则是 identity@imagename.jpg

**Comment by yzk0281** [comment]
Hi
我看你根目录下面最后三层:
layer {
  name: "norm2"
  type: "Python"
  bottom: "fc9"
  top: "norm2"
  python_param {
    module: "tripletloss.tripletlayer"
    layer: "Norm2Layer"
  }
}
layer {
  name: "triplet_select"
  type: "Python"
  bottom: "norm2"
  bottom: "labels"
  top: "archor"
  top: "positive"
  top: "negative"
  python_param {
    module: "tripletloss.tripletselectlayer"
    layer: "TripletSelectLayer"
    param_str: "'triplet': 30"
  }
}
layer {
  name: "tripletloss"
  type: "Python"
  bottom: "archor"
  bottom: "positive"
  bottom: "negative"
  top: "loss"
  python_param {
    module: "tripletloss.tripletlosslayer"
    layer: "TripletLayer"
    param_str: "'margin': 0.2"
  }
}
倒数第三层到底是：
layer {
  name: "norm2"
  type: "Python"
  bottom: "fc9"
  top: "norm2"
  python_param {
    module: "tripletloss.tripletlayer"
    layer: "Norm2Layer"
  }
}
还是 
layer {
  name: "norm2"
  type: "Python"
  bottom: "fc9"
  top: "norm2"
  python_param {
    module: "tripletloss.norm2layer"
    layer: "Norm2Layer"
  }
}

**Comment by pinguo-luhaofang** [comment]
谢谢提醒，确实是写错了，是norm2layer

**Comment by yzk0281** [comment]
train.py 老报错:
blob.py 31行
‘NoneType’ object has no attribute 'astype'
打印出来的im文件是有的  我在python命令行里单独
im = cv2.imread(..)
im.astype(..)
是成功的
有什么建议吗？

**Comment by pinguo-luhaofang** [comment]
‘NoneType’ object has no attribute 'astype'
请检查一下是不是你的文件有空行，或者是有文件没有找到
